File: Programmers/pro_72411_메뉴리뉴얼.py
from collections import defaultdict
from itertools import combinations

def solution(orders, course):
    menu_dict = defaultdict(int)
    max_cnt =defaultdict(int)
    for order in orders:
        for i in range(1,len(order)+1):
            comb = list(combinations(order,i))
            for elem in comb:
                elem = ''.join(sorted(elem))
                for k in course:
                    if len(elem)==k:
                        menu_dict[elem]+=1
                        max_cnt[str(k)]=max(max_cnt[str(k)],menu_dict[elem])
                        
    answer = []
    for elem, cnt in menu_dict.items():
        l = len(elem)
        if(l<2 or cnt<2):
            continue
        maxN = max_cnt[str(l)]
        if(cnt==maxN):
            answer.append(elem)
    return sorted(answer)

File: Programmers/pro_72411_메뉴 리뉴얼.js
function combination(arr, selectNum) {
    const result = [];
    if (selectNum === 1) return arr.map((v) => [v]);
    arr.forEach((v, idx, arr) => {
        const fixed = v;
        const restArr = arr.slice(idx + 1);
        const combinationArr = combination(restArr, selectNum - 1);
        const combineFix = combinationArr.map((v) => [fixed, ...v]);
        result.push(...combineFix);
    });
    return result;
}

function solution(orders, course) {
    const course_dict = []
    const max_dict = []
    for (let i = 0; i < orders.length; i++) {
        let order = [...orders[i]].sort()
        for (let j = 1; j <= order.length; j++) {
            if (course.indexOf(j) != -1) {
                let comb = combination(order, j)
                for (let k = 0; k < comb.length; k++) {
                    let elem = comb[k].join("")
                    course_dict[elem] ? course_dict[elem] += 1 : course_dict[elem] = 1
                    max_dict[j] ? max_dict[j] = Math.max(course_dict[elem], max_dict[j]) : max_dict[j] = course_dict[elem]
                }
            }

        }
    }
    let answer = [];
    for (let i in course_dict) {
        let cnt = course_dict[i]
        if (cnt >= 2 && cnt == max_dict[i.length])
            answer.push(i)
    }
    return answer.sort();
}